### PR TITLE
Update to oneDNN v3.0

### PR DESCRIPTION
**Pull request by eralmual**

### CODE REVIEW

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_conv.cc [hunk 1]
```diff
@@ -21,13 +21,13 @@ void DnnlConv::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode& node) {
 
   auto conv_src_mem = sp.GetMemory(node.Input(IN_X));
   auto src_md = conv_src_mem.get_desc();
-  src_md.data.format_kind = dnnl_format_kind_t::dnnl_format_kind_any;
-  auto src_dims = conv_src_mem.get_desc().dims();
+  src_md = dnnl::memory::desc(src_md.get_dims(), src_md.get_data_type(), dnnl::memory::format_tag::any);
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_conv.cc:24]: in the original code this memory descriptor was forced to be format_tag::any so that oneDNN chould choose the most optimal input memory shapes. This had a 10x impact on performance. Have you checked this still give the correct performance?

Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_conv.cc:24]: I'm gathering results, but should we expect any difference? the new src_md is exactly the same as before, the only change is that since the object is now opaque we cannot directly change it's format as we did before, so instead we just create a new one with the same dimensions, data type and desired tag (_dnnl::memory::format_tag::any_)

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_dynamicquantizelinear.cc [hunk 2]
```diff
@@ -85,43 +88,48 @@ void DnnlDynamicQuantizeLinear::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlN
...
-    div_saturate_round.append_eltwise(1.0f, dnnl::algorithm::eltwise_round, 0.0f, 0.0f);
+    div_saturate_round.append_eltwise(dnnl::algorithm::eltwise_round, 0.0f, 0.0f);
     zp_prim_attr.set_post_ops(div_saturate_round);
   }
+  // Set the value to scale DNNL_ARG_SRC_0 with mask 0
+  zp_prim_attr.set_scales_mask(DNNL_ARG_SRC_0, 0);
+  // Create the memory object related to the scale
+  auto scale_mem = dnnl::memory({{1}, dnnl::memory::data_type::f32, {1}}, eng);
+  // Write the alpha value into the memory object
+  dnnl_util::WriteToDnnlMemory<float>(scale_mem, {-1.0f});
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_dynamicquantizelinear.cc]: Is WriteToDnnlMemory a new function?

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_elementwise.cc [hunk 3]
```diff
@@ -35,17 +35,25 @@ void DnnlElementwise::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode& node)
       }
       break;
     }
+    case dnnl::algorithm::eltwise_soft_relu: {
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_elementwise.cc:38]: are we ever going to get eltwise_soft_relu algorithm for anything other than Softplus?

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_gemm.cc [hunk 4]
```diff
@@ -92,14 +93,17 @@ void DnnlGemm::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode& node) {
...
+  auto alpha_mem = dnnl::memory({{1}, dnnl::memory::data_type::f32, {1}}, eng);
+  // Write the alpha value into the memory object 
+  dnnl_util::WriteToDnnlMemory<float>(alpha_mem, {alpha});
 
   auto matmul_dst_md = dnnl::memory::desc(output_shape, node.Output(OUT_Y).Type(), {N, 1});
 
-  auto matmul_d = dnnl::matmul::desc(a_md, b_md, matmul_dst_md);
   dnnl::matmul::primitive_desc matmul_pd;
-  matmul_pd = dnnl::matmul::primitive_desc(matmul_d, matmul_attr, eng);
+  matmul_pd = dnnl::matmul::primitive_desc(eng, a_md, b_md, matmul_dst_md, matmul_attr);
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_gemm.cc:106]: Just checking, did oneDNN 3.0 added a attribute to the matmul. There is not default matmul that can be called without the new attribute and we must set the scales mask attribute since we are using the alpha attribute correct?

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_layernorm.cc [hunk 5]
```diff
@@ -174,10 +172,8 @@ void DnnlLayerNorm::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode& node) {
 
   // Get epsilon to avoid zero division
   float epsilon = GetEpsilon(node);
-  // Operation desciptor
-  auto lnorm_desc = dnnl::layer_normalization_forward::desc(prop_kind, src_md, epsilon, op_flags);
   // Primitive desciptor
-  auto lnorm_pd = dnnl::layer_normalization_forward::primitive_desc(lnorm_desc, dnnl_engine);
+  auto lnorm_pd = dnnl::layer_normalization_forward::primitive_desc(dnnl_engine, prop_kind, src_md, src_md, epsilon, op_flags);
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_layernorm.cc:176]: in the original layer_normalization desc src_md is only listed one time yet here is it listed twice.  Is that correct?

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_lrn.cc [hunk 6]
```diff
@@ -24,17 +24,35 @@ void DnnlLrn::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode& node) {
...
+  auto workspace_mem = dnnl::memory(lrn_pd.workspace_desc(), dnnl_engine);
+
+  sp.AddPrimitive(lrn_op, {{DNNL_ARG_SRC, lrn_src_mem},
+                            {DNNL_ARG_WORKSPACE, workspace_mem},
+                            {DNNL_ARG_DST, lrn_dst_mem}});
+#else
   sp.AddPrimitive(lrn_op, {{DNNL_ARG_SRC, lrn_src_mem},
-                       {DNNL_ARG_DST, lrn_dst_mem}});
+                           {DNNL_ARG_DST, lrn_dst_mem}});
+#endif  // ENABLE_TRAINING
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_lrn.cc:55]: Do we have unit tests that validate lrn training?

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_elementwise.cc (hunk 3 above)
Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_elementwise.cc:38]: From what I saw in the code, it's also used for sigmoid, that's why we check if the OpType is Softplus

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_dynamicquantizelinear.cc (hunk 2 above)
Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_dynamicquantizelinear.cc]: Yes, now the scales are passed as memory objects so we need to write the vector into the dnnl data handle, that's what this function does

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_gemm.cc (hunk 4 above)
Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_gemm.cc:106]: Yes, another thing to note is that some primitives don't support scaling the output (as we did before with _matmul_attr.set_output_scales(0, alphaScale);_ ) so in this case we scaled one input, this has the same result because of the operation

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_layernorm.cc (hunk 5 above)
Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_layernorm.cc:176]: Yes, previously most of the primitive descriptors didn't require a dst_md to generate the descriptor, now all primitive descriptors require a memory descriptor for each input and output they have, since this primitive is going to be executed in place the src_md is the same as the dst_md

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_lrn.cc (hunk 6 above)
Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_lrn.cc:55]: Let me check

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_pow.cc [hunk 7]
```diff
@@ -45,8 +45,12 @@ void DnnlPow::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode& node) {
   }
 
   // DNNL eltwise_pow is defined as alpha*x^beta. We don't use alpha so it is hard coded to 1.0
-  dnnl::eltwise_forward::desc elementwise_desc(dnnl::prop_kind::forward_inference, dnnl::algorithm::eltwise_pow, src_md, 1.0, beta);
-  dnnl::eltwise_forward::primitive_desc elementwise_pd(elementwise_desc, dnnl_engine);
+  dnnl::eltwise_forward::primitive_desc elementwise_pd(dnnl_engine, dnnl::prop_kind::forward_inference,
+                                                      dnnl::algorithm::eltwise_pow, src_md,
+                                                      dnnl::memory::desc(src_md.get_dims(),
+                                                                         src_md.get_data_type(),
+                                                                         dnnl::memory::format_tag::any),
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_pow.cc]: Could we make the memory desc on a separate line. I think it would improve readability.

i.e.

    // I don't know what the best name of this variable is I am just using src2_md here.
    // If you have a better name like dst_md please use it
    auto src2_md = dnnl::memory::desc(src_md.get_dims(),
                                      src_md.get_data_type(),
                                      dnnl::memory::format_tag::any);
    dnnl::eltwise_forward::primitive_desc elementwise_pd(dnnl_engine, dnnl::prop_kind::forward_inference,
                                                         dnnl::algorithm::eltwise_pow, src_md, src2_md,
                                                         1.0, beta);

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_qattention.cc [hunk 8]
```diff
@@ -189,19 +188,16 @@ void DnnlQAttention::CreatePrimitive(DnnlSubgraphPrimitive& sp, DnnlNode& node)
...
-
     dnnl::primitive_attr binary_attr;
     //scale source 0, matmul output
     if (total_scale_mem) {
-      binary_attr.set_scales(DNNL_ARG_SRC_0, 0, {DNNL_RUNTIME_F32_VAL});
+      binary_attr.set_scales_mask(DNNL_ARG_SRC_0, 0);
     }
 
-    auto binary_pd = dnnl::binary::primitive_desc(binary_d, binary_attr, eng);
+    auto binary_pd = dnnl::binary::primitive_desc(eng, dnnl::algorithm::binary_add, QKV_desc, bias_md, QKV_desc, binary_attr);
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_qattention.cc:200]: Something here looks/feel wrong but I can not pin point why. Can you double check this is correct? Sorry that I can not give an exact thing to check.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_subgraph_primitive.cc [hunk 9]
```diff
@@ -64,7 +64,7 @@
 * Then rebuild and run the code.
 * This is a developer only solution to investigating contents of OneDNN's tensors.
 */
-#define DNNL_TENSOR_PRINT_MEMORY 0
+#define DNNL_TENSOR_PRINT_MEMORY 1
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_subgraph_primitive.cc]: restore to print_memory 0 before doing public PR.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_subgraph_primitive.cc [hunk 10]
```diff
@@ -666,23 +664,20 @@ onnxruntime::common::Status DnnlSubgraphPrimitive::Predict(const std::unordered_
...
+      outputs_.at(output.first).set_data_handle(output.second.buffer);
     }
   }
 
   // reshapes (eg, unsqueeze)
   // it is safe to set data handle because all external data handles have been set and onednn managed memory data handles will not change
   for (auto& reshape_pair : reshapes_) {
-    reshape_pair.second.set_data_handle(reshape_pair.first.get_data_handle(),stream);
-    stream.wait();
+    reshape_pair.second.set_data_handle(reshape_pair.first.get_data_handle());
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_subgraph_primitive.cc:680]: Why did you remove the 3 instances of stream.wait()? is it no longer supported? I belive they were put there to make sure the outputs_, inputs_, and reshapes were guarantied.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_util.h [hunk 11]
```diff
@@ -10,8 +10,18 @@ namespace onnxruntime {
...
+
 dnnl::algorithm OrtOperatorToDnnlAlgorithm(std::string op);
+
+template <typename T>
+inline void WriteToDnnlMemory(dnnl::memory& mem, std::vector<T> values) {
+  // Get memory data handle
+  auto data_handle = static_cast<T*>(mem.get_data_handle());
+  // Copy data from the vector into the pointer
+  std::copy(values.begin(), values.end(), data_handle);
+}
```
Comment (georgen117) [inline onnxruntime/core/providers/dnnl/subgraph/dnnl_util.h]: Nice!

**Comment by georgen117** [comment]
I think the code change is great. Almost everything I did was ask some questions. I think I only suggested a single change.

The other thing that you may want to decided is if some of the line lengths should be broken across a few lines more lines. Some are long enough that they are auto wrapping when doing the side by side comparison. I personally try to make it possible to do side-by-side comparison without random line wraps. I am aware this is a personal choice but I would encourage browsing the changes and deciding if an added line wrap would improve readability in a side-by-side compare.

File: onnxruntime/core/providers/dnnl/subgraph/dnnl_subgraph_primitive.cc (hunk 10 above)
Comment (eralmual) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_subgraph_primitive.cc:680]: they are no longer needed, the _set_data_handle_ method does not need a dnnl stream

Comment (georgen117) [reply onnxruntime/core/providers/dnnl/subgraph/dnnl_subgraph_primitive.cc:680]: thanks I missed that the stream was removed from the set_data_handle function.
